framework/src/main/java/org/checkerframework/common/wholeprograminference/SceneToStubWriter.java

```diff
@@ -0,0 +1,439 @@
+     *
+     * <p>Takes two extra parameters beyond the AField: {@code fieldName} is the name to use for the
+     * declaration in the stub file. This doesn't matter for parameters, but must be correct for
+     * fields. {@code className} is the name of the enclosing class. This is only used for printing
+     * the type of an explicit receiver parameter (i.e. a parameter named "this").
+     */
+    private void printAField(AField aField, String fieldName, String className) {
+        String basetype;
+        if ("this".equals(fieldName)) {
+            basetype = className;
+        } else if (basetypes.containsKey(aField.type.description.toString())) {
+            basetype = basetypes.get(aField.type.description.toString()).toString();
+        } else {
+            basetype = "TODOTYPE";
+        }
+
+        // Annotations on arrays should be printed after the component type.
+        // Don't even bother checking for annotations on component types,
+        // since we can't infer them (yet?).
+        if (basetype.contains("[")) {
+            String component = basetype.substring(0, basetype.lastIndexOf('['));
+            printArrayComponentTypeAnnotation(aField.type);
+            printWriter.print(component);
+            printWriter.print(" ");
+            basetype = "[]";
+        }
+        printAnnotations(aField.type.tlAnnotationsHere);
+        printWriter.print(" ");
+        printWriter.print(basetype);
+        printWriter.print(" ");
+        printWriter.print(fieldName);
+    }
+
+    /**
+     * {@code DefCollector} is a facility in the annotation file utilities for determining which
+     * annotations are used in a given AScene, and then writing out their definitions into the index
+     * file. Here, we abuse that construct to write out the proper import statements into a stub
+     * file.
+     */
+    private class ImportDefCollector extends DefCollector {
+
+        /**
+         * Constructs a new ImportDefCollector, which will run on the given AScene when its {@code
+         * visit} method is called.
+         */
+        ImportDefCollector(AScene scene) throws DefException {
+            super(scene);
+        }
+
+        /**
+         * Write an import statement for a given AnnotationDef. Don't do anything fancy, and assume
+         * that DefCollector did its job so that this will only be called once on each annotation.
+         */
+        @Override
+        protected void visitAnnotationDef(AnnotationDef d) {
+            if (!d.name.contains("+")) {
+                printWriter.println("import " + d.name + ";");
+            }
+        }
+    }
+
+    /**
+     * When an inner class is present in an AScene, its name is something like "Outer$Inner".
+     * Writing a stub file with that name would be useless to the stub parser, which expects inner
+     * classes to be properly nested.
+     *
+     * <p>For a given class, this prints out the hierarchy of outer classes, and returns the number
```

Comment:
So this prints each class listed in `classname`?

Reply:
Ping

Reply:
Yes. (sorry for missing this comment earlier)

I'll update this comment.